### ISSUE TITLE
CachingLoaderMixin requires a cache size greater than zero

### ISSUE THREAD

**Issue by chaseastewart** [jg-rp/liquid#148]
The [CachingLoaderMixin](https://github.com/jg-rp/liquid/blob/ea0d6d817dcd82a844588ce75d31947b16ce91a4/liquid/builtin/loaders/mixins.py#L92) requires cache size to be greater than zero due to [LRUCache](https://github.com/jg-rp/liquid/blob/ea0d6d817dcd82a844588ce75d31947b16ce91a4/liquid/utils/cache.py#L148) not accounting for a capacity of zero. For now I've had to ensure the cache size is at least 1 in my [loader](https://github.com/chaseastewart/fhir-converter/blob/3ccbdad3f2701e93a02a5d3e8615b760192b82a8/fhir_converter/loaders.py#L70). Ideally the mixin would work if cache size is >= 0 to allow loaders to support caching to be disabled similar to the enviroment [cache](https://github.com/jg-rp/liquid/blob/ea0d6d817dcd82a844588ce75d31947b16ce91a4/liquid/environment.py#L222).

**Comment by chaseastewart** [comment]
@jg-rp The way I handled this in my previous version was to set cache to an empty dict, use a boolean to know if caching is enabled though this was really a work around LRUCache.

**Comment by jg-rp** [comment]
It's temping - to avoid the extra overhead when checking the cache - to make `LRUCache` fail with a `ValueError` at construction time when given a zero or negative capacity, forcing the caller to pick a more appropriate loader. Maybe also provide convenience class methods on `CachingFileSystemLoader` and `CachingChoiceLoader`.

```python
class CachingFileSystemLoader(CachingLoaderMixin, FileExtensionLoader):

    # ..

    @classmethod
    def make_loader(
        cls,
        search_path: Union[str, Path, Iterable[Union[str, Path]]],
        encoding: str = "utf-8",
        ext: str = ".liquid",
        *,
        auto_reload: bool = True,
        namespace_key: str = "",
        cache_size: int = 300,
    ) -> BaseLoader:
        try:
            return cls(
                search_path=search_path,
                encoding=encoding,
                ext=ext,
                auto_reload=auto_reload,
                namespace_key=namespace_key,
                cache_size=cache_size,
            )
        except ValueError:
            return FileExtensionLoader(
                search_path=search_path, encoding=encoding, ext=ext
            )
```

Or maybe make it transparent using `__new__`.

```python
class CachingFileSystemLoader(CachingLoaderMixin, FileExtensionLoader):

    # ..

    def __new__(
        cls,
        search_path: Union[str, Path, Iterable[Union[str, Path]]],
        encoding: str = "utf-8",
        ext: str = ".liquid",
        *,
        cache_size: int = 300,
    ) -> BaseLoader:
        if cache_size < 1:
            return FileExtensionLoader(
                search_path=search_path,
                encoding=encoding,
                ext=ext,
            )
        return super().__new__(cls)
```

The explicit `make_loader()` approach will probably be less surprising and be easier to override should anyone want to subclass a caching loader.

**Comment by chaseastewart** [comment]
@jg-rp My main concern is maintaining compatibility with FHIR-Converter's dynamic naming convention. The names inside a template do not match the actual file names. See [_resolve_template_name](https://github.com/chaseastewart/fhir-converter/blob/3ccbdad3f2701e93a02a5d3e8615b760192b82a8/fhir_converter/loaders.py#L87). To achieve this I'm currently overriding the get_source / get_source_with_context methods. I'd have to override the cached and non cached loaders. If there was away to wire in a TemplateNameResolver, the actual loader implementation wouldn't matter.

**Comment by jg-rp** [comment]
> My main concern is maintaining compatibility with FHIR-Converter's dynamic naming convention.

Ah, I see. Would moving `_resolve_temple_name` (or similar) to a `liquid.Environment` subclass work for you? This would rewrite template names before they reach the template loader.

```python
from liquid import Environment

# ...

class FHIRConverterEnv(Environment):

    def _template_name_hook(self, template_name: str) -> str:
        template_path = Path(template_name)
        parts = template_path.parts
        if len(parts) > 1:
            tail = parts[-1]
            if not tail.endswith(".json") and not tail.startswith("_"):
                template_path = template_path.with_name("_" + tail)
        return str(template_path)

    def get_template(
        self,
        name: str,
        globals: Optional[Mapping[str, object]] = None,
    ) -> BoundTemplate:
        return super().get_template(self._template_name_hook(name), globals)

    def get_template_with_context(
        self,
        context: "Context",
        name: str,
        **kwargs: str,
    ) -> BoundTemplate:
        return super().get_template_with_context(context, self._template_name_hook(name), **kwargs)
```

Changing the template loader API is already on [my list](https://github.com/jg-rp/liquid/discussions/137) for Python Liquid version 2. Having so many `get_template*`, `get_source*` and `load*` methods is pretty awful.

**Comment by chaseastewart** [comment]
@jg-rp This works. Are you going to move forward with make_loader? Thanks.

**Comment by jg-rp** [comment]
@chaseastewart Yes (#149), although I've gone with `make_choice_loader()` and `make_file_system_loader()`, both standalone factory functions instead of class methods.

**Comment by jg-rp** [comment]
`make_choice_loader()` and `make_file_system_loader()` are available in Python Liquid version 1.12.0, now released.

Example usage:

```python
from liquid import Environment
from liquid import FileSystemLoader
from liquid import PackageLoader
from liquid import make_choice_loader

template_loaders = [
    FileSystemLoader("/path/to/templates"),
    PackageLoader("some_package", package_path="templates"),
]

loader = make_choice_loader(template_loaders)
print(loader.__class__.__name__)  # CachingChoiceLoader

loader = make_choice_loader(template_loaders, cache_size=0)
print(loader.__class__.__name__)  # ChoiceLoader

env = Environment(loader=loader)
```
